src/txmempool.cpp

```diff
@@ -429,5 +438,159 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
+                                 CAmount& nFeesReserved, CAmount& nFeesRemoved) {
+    std::set<uint256> protect;
+    BOOST_FOREACH(const CTxIn& in, toadd.GetTx().vin) {
+        protect.insert(in.prevout.hash);
+    }
+
+    size_t incUsage = GuessDynamicMemoryUsage(toadd);
+    size_t expsize = DynamicMemoryUsage() + incUsage; // Track the expected resulting memory usage of the mempool.
+    if (expsize > sizelimit) {
+        size_t sizeToTrim = std::min(expsize - sizelimit, incUsage);
+        return TrimMempool(sizeToTrim, protect, nFeesReserved, toadd.GetTxSize(), toadd.GetFee(), stage, nFeesRemoved, true);
+    }
+    else
+        return true;
+}
+
+bool CTxMemPool::SurplusTrim(int multiplier, CFeeRate minRelayRate, std::set<uint256>& stage, size_t usageToTrim) {
+    CFeeRate excessRate(multiplier * minRelayRate.GetFeePerK());
+    std::set<uint256> noprotect;
+    CAmount nFeesRemoved = 0;
+    size_t sizeToTrim = usageToTrim / 4;  //Conservatively assume we have transactions at least 1/4 the size of the mempool space they've taken
+    return TrimMempool(usageToTrim, noprotect, 0, sizeToTrim, excessRate.GetFee(sizeToTrim), stage, nFeesRemoved, false);
+}
+
+bool CTxMemPool::TrimMempool(size_t sizeToTrim, std::set<uint256> &protect, CAmount nFeesReserved, size_t sizeToUse, CAmount feeToUse,
+                             std::set<uint256>& stage, CAmount &nFeesRemoved, bool mustTrimAllSize) {
+    size_t usageRemoved = 0;
+    indexed_transaction_set::nth_index<1>::type::reverse_iterator it = mapTx.get<1>().rbegin();
+    int fails = 0; // Number of mempool transactions iterated over that were not included in the stage.
+    int itertotal = 0;
+    int iterextra = mustTrimAllSize ? 10 : 100; //Allow many more iterations to find large size during SurplusTrim
+    int iterperfail = mustTrimAllSize ? 5 : 10; //Try to follow more links during SurplusTrim
+    int failmax = 10; //Try no more than 10 starting transactions
+    uint64_t skipPattern = GetRand((uint64_t)1<<61);
+    // Iterate from lowest feerate to highest feerate in the mempool:
+    while (usageRemoved < sizeToTrim && it != mapTx.get<1>().rend()) {
+        //Randomly skip between 0 and 7 entries
+        unsigned int skipNum = std::min(fails,19);
+        for (unsigned int i = 0; i < ((uint64_t)7 & (skipPattern >> (3*skipNum))) && it != mapTx.get<1>().rend();i++,it++);
+        if (it == mapTx.get<1>().rend())
+            break;
+
+        const uint256& hash = it->GetTx().GetHash();
+        if (stage.count(hash)) {
+            // If the transaction is already staged for deletion, we know its descendants are already processed, so skip it.
+            it++;
+            continue;
+        }
+        if ((double)it->GetFee() * sizeToUse > (double)feeToUse * it->GetTxSize()) {
+            // If the transaction's feerate is worse than what we're looking for, we have processed everything in the mempool
```

Comment:
Agree the comment needs changing, but the check is still valuable.  No reason to continue checking further transactions at this point, even if we haven't used up all of our fails.